### PR TITLE
Cypress: implement SWAP mode for PSoC6

### PR DESCRIPTION
**Pull request by BohdanKovalchuk**
Signed-off by Bohdan Kovalchuk <bohdan.kovalchuk@linaro.org>

### CODE REVIEW

**Comment by BohdanKovalchuk** [comment]
@d3zd3z , @utzig please, review this draft code so we can progress on SWAP implementation for PSoC6

please, ignore // TODO and 1362 for now since the code isn't final

I'd appreciate your inputs on design

thanks

File: boot/bootutil/src/bootutil_misc.c [hunk 1]
```diff
@@ -195,11 +195,7 @@ uint32_t
 boot_status_off(const struct flash_area *fap)
 {
     uint32_t off_from_end;
-    uint8_t elem_sz;
-
-    elem_sz = flash_area_align(fap);
-
-    off_from_end = boot_trailer_sz(elem_sz);
+    off_from_end = boot_trailer_sz(BOOT_MAX_ALIGN);
```
Comment (utzig) [inline boot/bootutil/src/bootutil_misc.c:198]: This is not OK, we want to use the write size alignment for each particular MCU, this is not supposed to work on *your* MCU only.

File: boot/bootutil/src/swap_move.c [hunk 2]
```diff
@@ -140,7 +140,7 @@ swap_read_status_bytes(const struct flash_area *fap,
     found_idx = -1;
     /* skip erased sectors at the end */
     last_rc = 1;
-    write_sz = BOOT_WRITE_SZ(state);
+    write_sz = BOOT_MAX_ALIGN;
```
Comment (utzig) [inline boot/bootutil/src/swap_move.c:143]: Maybe you should consider making `BOOT_WRITE_SZ` return `BOOT_MAX_ALIGN` for your build?

File: boot/zephyr/CMakeLists.txt [hunk 3]
```diff
@@ -99,6 +99,7 @@ zephyr_library_sources(
   ${BOOT_DIR}/bootutil/src/image_rsa.c
   ${BOOT_DIR}/bootutil/src/image_ec256.c
   ${BOOT_DIR}/bootutil/src/image_ed25519.c
+  ${BOOT_DIR}/bootutil/src/bootutil_misc.c
```
Comment (utzig) [inline boot/zephyr/CMakeLists.txt:102]: This is not supposed to be part of this PR, is it? Maybe needs a rebase?

File: boot/bootutil/src/bootutil_misc.c [hunk 4]
```diff
@@ -457,6 +453,28 @@ boot_write_magic(const struct flash_area *fap)
...
 }
 
+
+/**
+ * Write trailer data; status bytes, swap_size, etc
+ *
+ * @returns 0 on success, != 0 on error.
+ */
+int
+cy_boot_write_trailer(const struct flash_area *fap, uint32_t off,
```
Comment (utzig) [inline boot/bootutil/src/bootutil_misc.c:463]: `cy_` prefixes are not acceptable here. Also what would happen if a reset happens while inside the `flash_area_write_byte` loop?

Comment (BohdanKovalchuk) [reply boot/bootutil/src/bootutil_misc.c:463]: yes, sure

I left it intentionally to highlight which API required to be implemented in different way

I'll change it to neutral as soon as we finalize the code itself

File: boot/bootutil/src/swap_move.c (hunk 2 above)
Comment (BohdanKovalchuk) [reply boot/bootutil/src/swap_move.c:143]: Is my understanding correct on following?

* BOOT_WRITE_SZ()  is a value of state->write_sz;
* state->write_sz is filled in with return value of flash_are_align();
which is, in context of SWAP status table write-operations, the same as BOOT_MAX_ALIGN;

So, is it still required different approaches - macro, API and #define name for the same value?

File: boot/zephyr/CMakeLists.txt (hunk 3 above)
Comment (BohdanKovalchuk) [reply boot/zephyr/CMakeLists.txt]: rebased

**Comment by BohdanKovalchuk** [comment]
- closing as unacceptable implementation

**Pull request closed without merging**